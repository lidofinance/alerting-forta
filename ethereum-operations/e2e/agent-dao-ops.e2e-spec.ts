import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-dao-ops e2e tests", () => {
  let runBlock: (blockHashOrNumber: string | number) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/dao-ops/agent-dao-ops")
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process block with low staking limit (10%)",
    async () => {
      let findings = await runBlock(16704075);
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with low staking limit (30%)",
    async () => {
      let findings = await runBlock(16704075);
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with huge buffered ETH amount and low deposit executor balance",
    async () => {
      let findings = await runBlock(17241550);
      findings.sort();
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with MEV Allow list: Super low relay count",
    async () => {
      let findings = await runBlock(15960625);
      findings.sort();
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  // todo: disabled due to changes in the event signatures.
  //  uncomment or remove after V2
  // it(
  //   "should process tx with paused deposits",
  //   async () => {
  //     const findings = await runTransaction(
  //       "0x878e2a2ff5018e680dfd37031ed8b5d1d7e2ec41a4bd640168f347bbedf31a7b"
  //     );
  //     expect(findings.at(0)).toMatchSnapshot();
  //   },
  //   TEST_TIMEOUT
  // );

  // it(
  //   "should process tx with unpaused deposits",
  //   async () => {
  //     const findings = await runTransaction(
  //       "0xd5ddc8c0e3e24b36ffac6e804894c5c09c012e9079a9d9df24f503f1e75e2f64"
  //     );
  //     expect(findings.at(0)).toMatchSnapshot();
  //   },
  //   TEST_TIMEOUT
  // );

  it(
    "should process tx with added guardian",
    async () => {
      let findings = await runTransaction(
        "0x60d9392de6c6ae3f8ca8003cce414fc420d705c9a1c8051b9869b0b870d2ebbe"
      );
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed guardian quorum",
    async () => {
      let findings = await runTransaction(
        "0x37ef7e9e71809be37143e3976bae7859268be3fac6728d69cd7b5d6d9cf8d24a"
      );
      findings.sort();
      expect(findings.at(4)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed max deposit",
    async () => {
      let findings = await runTransaction(
        "0x9c10852a83c77204f255705e581c21ebcc28c021dfaff4c02707a4cee1eedde2"
      );
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed min deposit block",
    async () => {
      let findings = await runTransaction(
        "0x9c10852a83c77204f255705e581c21ebcc28c021dfaff4c02707a4cee1eedde2"
      );
      findings.sort();
      expect(findings.at(1)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed owner of the contract",
    async () => {
      let findings = await runTransaction(
        "0x9c10852a83c77204f255705e581c21ebcc28c021dfaff4c02707a4cee1eedde2"
      );
      findings.sort();
      expect(findings.at(2)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed distribution fee",
    async () => {
      let findings = await runTransaction(
        "0xe61167aa87b2a7aa9bd68834bf703877d22315d6d765345ebf0135eb8c33c406"
      );
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with changed protocol contracts",
    async () => {
      let findings = await runTransaction(
        "0xdd76a4d06199eb017e322e2e152f88841d92488ef5e02809bac45c842244059e"
      );
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with EL rewards withdrawal limit set",
    async () => {
      let findings = await runTransaction(
        "0x3f126198f641194e96cf6be0db03efc6ae92a4e1586b5062aaf373f171f0c5d2"
      );
      findings.sort();
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with EL rewards vault set and staking changes",
    async () => {
      let findings = await runTransaction(
        "0x11a48020ae69cf08bd063f1fbc8ecf65bd057015aaa991bf507dbc598aadb68e"
      );
      findings.sort();
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with transferred ownership of Insurance fund",
    async () => {
      let findings = await runTransaction(
        "0x91c7c2f33faf3b5fb097138c1d49c1d4e83f99e1c3b346b3cad35a5928c03b3a"
      );
      findings.sort();
      expect(findings).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );
});
