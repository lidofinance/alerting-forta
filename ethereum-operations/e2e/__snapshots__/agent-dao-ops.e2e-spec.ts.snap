// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-dao-ops e2e tests should process block with MEV Allow list: Super low relay count 1`] = `
Array [
  Finding {
    "addresses": Array [],
    "alertId": "MEV-LOW-RELAY-COUNT",
    "description": "There are only 0 relays in the allowed list.",
    "labels": Array [],
    "metadata": Object {},
    "name": "üö® MEV Allow list: Super low relay count",
    "protocol": "ethereum",
    "severity": 4,
    "type": 4,
  },
]
`;

exports[`agent-dao-ops e2e tests should process block with huge buffered ETH amount and low deposit executor balance 1`] = `
Array [
  Finding {
    "addresses": Array [],
    "alertId": "LOW-DEPOSIT-EXECUTOR-BALANCE",
    "description": "Balance of deposit executor is 1.9232. This is extremely low! üò±",
    "labels": Array [],
    "metadata": Object {},
    "name": "‚ö†Ô∏è Low deposit executor balance",
    "protocol": "ethereum",
    "severity": 4,
    "type": 2,
  },
  Finding {
    "addresses": Array [],
    "alertId": "HUGE-BUFFERED-ETH",
    "description": "There are 41420.0866 buffered ETH in DAO for more than 1 hour(s)",
    "labels": Array [],
    "metadata": Object {},
    "name": "üö® Huge buffered ETH amount",
    "protocol": "ethereum",
    "severity": 4,
    "type": 3,
  },
]
`;

exports[`agent-dao-ops e2e tests should process block with low staking limit (10%) 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LOW-STAKING-LIMIT",
  "description": "Current staking limit is 7237.74 ETH this is lower than 10% of max staking limit 150000.00 ETH",
  "labels": Array [],
  "metadata": Object {},
  "name": "‚ö†Ô∏è Staking limit below 10%",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-dao-ops e2e tests should process block with low staking limit (30%) 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LOW-STAKING-LIMIT",
  "description": "Current staking limit is 7237.74 ETH this is lower than 30% of max staking limit 150000.00 ETH",
  "labels": Array [],
  "metadata": Object {},
  "name": "üìâ Staking limit below 30%",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-dao-ops e2e tests should process tx with EL rewards vault set and staking changes 1`] = `
Array [
  Finding {
    "addresses": Array [],
    "alertId": "LIDO-STAKING-LIMIT-SET",
    "description": "Staking limit was set with:
Max staking limit: 150000000000000000000000
Stake limit increase per block: 23437500000000000000",
    "labels": Array [],
    "metadata": Object {
      "args": "150000000000000000000000,23437500000000000000",
    },
    "name": "‚ö†Ô∏è Lido: Staking limit set",
    "protocol": "ethereum",
    "severity": 4,
    "type": 4,
  },
  Finding {
    "addresses": Array [],
    "alertId": "LIDO-STAKING-RESUMED",
    "description": "Staking was resumed!",
    "labels": Array [],
    "metadata": Object {
      "args": "",
    },
    "name": "‚úÖ Lido: Staking resumed",
    "protocol": "ethereum",
    "severity": 4,
    "type": 4,
  },
]
`;

exports[`agent-dao-ops e2e tests should process tx with EL rewards withdrawal limit set 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with added guardian 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with changed distribution fee 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with changed guardian quorum 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with changed max deposit 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with changed min deposit block 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with changed owner of the contract 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with changed protocol contracts 1`] = `undefined`;

exports[`agent-dao-ops e2e tests should process tx with received EL rewards 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "LIDO-EL-REWARDS-RECEIVED",
  "description": "Rewards amount: 142.21 ETH",
  "labels": Array [],
  "metadata": Object {
    "args": "142210964953261078203",
  },
  "name": "‚úÖ Lido: EL rewards received",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-dao-ops e2e tests should process tx with transferred ownership of Insurance fund 1`] = `
Array [
  Finding {
    "addresses": Array [],
    "alertId": "INS-FUND-OWNERSHIP-TRANSFERRED",
    "description": "Owner of the insurance fund was transferred from [0x0000000000000000000000000000000000000000](https://etherscan.io/address/0x0000000000000000000000000000000000000000) to [0xbD829522d4791b9660f59f5998faE451dACA4E1C](https://etherscan.io/address/0xbD829522d4791b9660f59f5998faE451dACA4E1C)",
    "labels": Array [],
    "metadata": Object {
      "args": "0x0000000000000000000000000000000000000000,0xbD829522d4791b9660f59f5998faE451dACA4E1C",
    },
    "name": "üö® Insurance fund: Ownership transferred",
    "protocol": "ethereum",
    "severity": 5,
    "type": 4,
  },
  Finding {
    "addresses": Array [],
    "alertId": "INS-FUND-OWNERSHIP-TRANSFERRED",
    "description": "Owner of the insurance fund was transferred from [0xbD829522d4791b9660f59f5998faE451dACA4E1C](https://etherscan.io/address/0xbD829522d4791b9660f59f5998faE451dACA4E1C) to [0x3e40D73EB977Dc6a537aF587D48316feE66E9C8c](https://etherscan.io/address/0x3e40D73EB977Dc6a537aF587D48316feE66E9C8c)",
    "labels": Array [],
    "metadata": Object {
      "args": "0xbD829522d4791b9660f59f5998faE451dACA4E1C,0x3e40D73EB977Dc6a537aF587D48316feE66E9C8c",
    },
    "name": "üö® Insurance fund: Ownership transferred",
    "protocol": "ethereum",
    "severity": 5,
    "type": 4,
  },
]
`;
